### PR TITLE
tracing: Add decorator propagate field to enable passing the operatio…

### LINKED ISSUE #9778: Remove x-envoy-decorator-operation response header?
Is there a way to remove the x-envoy-decorator-operation header from the response to the client? I'm not directly setting that header anywhere, I assume it is being set by Envoy using the decorator value. I've tried setting the response_headers_to_remove setting in the route_config but that has had no effect. Here is the route_config I am using:

>           
          route_config:
            name: xxxx
            response_headers_to_remove:
              - x-envoy-decorator-operation
            virtual_hosts:
            - name: xxxx
              domains:
              - "*"
              routes:
              - match:
                  prefix: "/"
                route:
                  timeout: 300s
                  idle_timeout:
                  cluster: xxxx
                decorator:
                  operation: xxxx
          http_filters:
          - name: envoy.router
            config:
              suppress_envoy_headers: true

### PR DESCRIPTION
**Pull request by objectiser**
…n via headers to be disabled

Signed-off-by: Gary Brown <gary@brownuk.com>

Fixes #9778 

Description: Add a field to enable the decorator operation propagation (via http headers) to be disabled.
Risk Level: low
Testing: Added test to check whether the header is included in requests and responses. Existing behaviour is preserved.
Docs Changes: Additional `propagate` field to be documented.
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -933,6 +933,8 @@ void ConnectionManagerImpl::ActiveStream::traceRequest() {
   if (hasCachedRoute() && cached_route_.value()->decorator()) {
     cached_route_.value()->decorator()->apply(*active_span_);
 
+    decorated_propagate_ = cached_route_.value()->decorator()->getPropagate();
```
Comment (dio) [inline source/common/http/conn_manager_impl.cc]: Super nit, hold local ref of `cached_route_.value()->decorator()`?

File: include/envoy/router/router.h [hunk 2]
```diff
@@ -816,6 +816,13 @@ class Decorator {
    * @return the operation name
    */
   virtual const std::string& getOperation() const PURE;
+
+  /**
+   * This method returns whether the decorator information
+   * should be propagated to other services.
+   * @return whether to propagate
+   */
+  virtual bool getPropagate() const PURE;
```
Comment (lizan) [inline include/envoy/router/router.h]: nit: prefer `propagate` for method name.

File: source/common/http/conn_manager_impl.h [hunk 3]
```diff
@@ -683,6 +683,7 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     uint32_t buffer_limit_{0};
     uint32_t high_watermark_count_{0};
     const std::string* decorated_operation_{nullptr};
+    bool decorated_propagate_ = true;
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h]: While you are in here, do you mind moving all of these booleans up to the `State` struct above and using the bit-field where possible to save some space?

**Comment by objectiser** [comment]
@mattklein123 Build seems to be failing following merge of https://github.com/envoyproxy/envoy/pull/9918. Any ideas - not sure why these tests didn't fail with that PR. Will try taking another look early next week, but any thoughts appreciated.

File: source/common/http/conn_manager_impl.h [hunk 4]
```diff
@@ -615,6 +617,18 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
       // Used to track which filter is the latest filter that has received data.
       ActiveStreamEncoderFilter* latest_data_encoding_filter_{};
       ActiveStreamDecoderFilter* latest_data_decoding_filter_{};
+
+      bool decorated_propagate_ : 1;
+      // By default, we will assume there are no 100-Continue headers. If encode100ContinueHeaders
+      // is ever called, this is set to true so commonContinue resumes processing the 100-Continue.
+      bool has_continue_headers_ : 1;
+      bool is_head_request_ : 1;
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h]: These need to be packed above with the other boolsn to get the wanted effect. Can you move them up there?

**Comment by mattklein123** [comment]
> @mattklein123 Build seems to be failing following merge of #9918. Any ideas - not sure why these tests didn't fail with that PR. Will try taking another look early next week, but any thoughts appreciated.

The new tests you added need to be modified similarly to how the tests in my PR were modified. Take a look at the diff in my PR and compare to your tests and it should be pretty obvious what to change. Sorry for the trouble.

**Comment by mattklein123** [comment]
Sorry looks like there is a merge conflict now. :(

/wait

**Comment by objectiser** [comment]
Hopefully this should fix everything now :)

**Pull request merged**